[WIP] Added -swift-version support for 4.2

tkremenek opened the pull request.

tkremenek committed:
Add “4.2” to ValidEffectiveVersions.

tkremenek committed:
Allow ‘getEffectiveLanguageVersion’ to take a minor component for 4.2

Further, when ‘4’ is specified on its own for the language
version without a minor component, assume ‘4.1’ as the
language version.

tkremenek commented:
@jrose-apple This is incomplete.  I just wanted to make sure this looks on the right track as far as specifying the version mode to "4.1" with --swift-version 4.

jrose-apple commented on lib/Basic/Version.cpp:
```diff
@@ -326,6 +327,9 @@ Optional<Version> Version::getEffectiveLanguageVersion() const {
+    // Version '4' on its own implies '4.1'.
+    if (size() == 1)
+      return Version{4, 1};
     return Version::getCurrentLanguageVersion();
```
I feel like this should at least *check* for 4.0 and 4.1.

tkremenek replied:
Check in what way?  The driver will disallow passing "4.0" and "4.1" explicitly.  The only thing the driver will allow is passing "4.2".  We could add an assert that, if there is a second component, that it is "2".  WDYT?

jrose-apple replied:
Oh, right, the allowed-language-versions check. Okay, yes, that assert would be good enough for me.

jrose-apple replied:
Ah, it shouldn't be 4.1, though. It needs to be higher than any 4.1 point updates we release, right? So 4.1.99 or something.

tkremenek replied:
I spoke wrong.  The allowed-language-versions check is defined in terms of `getEffectiveLanguageVersion`.

jrose-apple reviewed (commented):
Looks fine so far. If there are any problems in this whole endeavour I'd expect it to be in the importer.

tkremenek committed:
Teach ClangImporter to handle effective Swift version with minor release.

Needed to support Swift 4.2.

tkremenek commented:
@jrose-apple I added the necessary logic to teach the ClangImporter about 4.2 and to handle minor versions.  It's not very pretty.  Most of it is OK, except:

* It feels like a form of DRY violations to special case "4.2" in a couple places.
* The logic in `ImportName.cpp` feels especially brittle.  The entire implementation of `ImportNameVersion` assumed major versions were the only thing the importer needed to care about.  I came up with a solution that fits in the existing implementation fine, but feels brittle from a maintainability perspective.

tkremenek commented:
@swift-ci smoke test

jrose-apple commented on lib/ClangImporter/ImportName.h:
```diff
@@ -56,12 +56,35 @@ class ImportNameVersion : public RelationalOperationsBase<ImportNameVersion> {
+    // We encode the 'rawValue' as just major version numbers with the
+    // exception of '4.2', which is a special minor version that can impact
+    // importing of names.  We treat that with a rawValue of 5, and treat
+    // all major values of 5 or higher as being rawValue = majorversion + 1.
```
Since we don't distinguish between Swift 1 and Swift 2, what do you think about moving the raw values down instead of up? It's a trick that will only work once, but even so.

jrose-apple commented on test/APINotes/Inputs/custom-frameworks/APINotesFrameworkTest.framework/Headers/APINotesFrameworkTest.apinotes:
```diff
@@ -259,7 +259,7 @@ SwiftVersions:
       - Name: multiVersionedGlobal45Both
         SwiftName: multiVersionedGlobal45Both_5
-  - Version: 4 # Versions are deliberately ordered as "3, 5, 4" to catch bugs.
+  - Version: 4 # Versions are deliberately ordered as "3, 5, 4.2, 4" to catch bugs.
```
Can you add a separate test that shows that "Version: 4.2" with no other versions specified also affects `-swift-version 4` and `-swift-version 3`?, and that "Version: 4" with no other versions does *not* affect `-swift-version 4.2`? Or maybe that's in here already; the original 3-4-5 tests did do things like that.

tkremenek replied:
@jrose-apple I added this in the latest commit.

jrose-apple reviewed (commented):
Looking good so far!

jrose-apple commented:
I do agree about the 4.2 being repeated. I'll see if I can think of a better way as well.

tkremenek committed:
Add APINotes test for just “Version 4.2”.

This tests that a “Version 4.2” change impacts
-swift-version 3 and 4, but not 5.

tkremenek commented:
@swift-ci smoke test

tkremenek commented:
@jrose-apple what do you think about landing this as is, and further refinements to `ImportNameVersion` (if any) in follow up?

jrose-apple reviewed (approved):
Sure, that sounds reasonable to me.
